### PR TITLE
Implement conversions between GlweCiphertextVectorView64 and CudaGlweCiphertextVector64

### PR DESCRIPTION
**Pull request by pdroalves**
### Description
Implements: 

- Implements GlweCiphertextVectorView64 and the related engines for creation, consumption, and serialization
- Conversion from GlweCiphertextVectorView64 to CudaGlweCiphertextVector64,
- Discarding conversion from CudaGlweCiphertextVector64 to GlweCiphertextVectorMutView64

### Checklist 

(Use '[x]' to check the checkboxes, or submit the PR and then click the checkboxes)

* [x] Tests for the changes have been added (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)
* [x] The PR description links to the related issue (to link an issue, use '#XXX'.)
* [x] The tests on AWS have been launched and are successful (apply the `aws_test` to the PR to launch the tests on AWS)
* [x] The draft release description has been updated
* [x] Check for breaking changes and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Comment by pdroalves** [comment]
@slab-ci gpu_test

**Comment by pdroalves** [comment]
@slab-ci cpu_test

**Comment by pdroalves** [comment]
@slab-ci gpu_test

**Comment by pdroalves** [comment]
@slab-ci cpu_test

**Comment by pdroalves** [comment]
@slab-ci cpu_test

**Comment by pdroalves** [comment]
@slab-ci cpu_test

**Comment by pdroalves** [comment]
@slab-ci cpu_test

**Comment by pdroalves** [comment]
@slab-ci cpu_test

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

File: concrete-core/src/backends/cuda/implementation/engines/cuda_engine/glwe_ciphertext_vector_conversion.rs [hunk 1]
```diff
@@ -344,3 +347,208 @@ impl GlweCiphertextVectorConversionEngine<CudaGlweCiphertextVector64, GlweCipher
         ))
     }
 }
+
```
Comment (IceTDrinker) [inline concrete-core/src/backends/cuda/implementation/engines/cuda_engine/glwe_ciphertext_vector_conversion.rs:350]: no 32 bits ?

File: concrete-core/src/backends/cuda/implementation/engines/cuda_engine/glwe_ciphertext_vector_conversion.rs [hunk 2]
```diff
@@ -344,3 +347,208 @@ impl GlweCiphertextVectorConversionEngine<CudaGlweCiphertextVector64, GlweCipher
...
+        let data_per_gpu = input.glwe_ciphertext_count().0
+            * input.glwe_dimension().to_glwe_size().0
+            * input.polynomial_size().0;
+        for gpu_index in 0..self.get_number_of_gpus() {
+            let stream = &self.streams[gpu_index];
+            let mut vec = stream.malloc::<u64>(data_per_gpu as u32);
+            let input_slice = input.0.as_tensor().as_slice();
+            stream.copy_to_gpu::<u64>(&mut vec, input_slice);
+            vecs.push(vec);
+        }
```
Comment (IceTDrinker) [inline concrete-core/src/backends/cuda/implementation/engines/cuda_engine/glwe_ciphertext_vector_conversion.rs:437]: no remainder management here ?

File: concrete-core/src/backends/cuda/implementation/engines/cuda_engine/glwe_ciphertext_vector_conversion.rs [hunk 3]
```diff
@@ -344,3 +347,208 @@ impl GlweCiphertextVectorConversionEngine<CudaGlweCiphertextVector64, GlweCipher
...
```
Comment (IceTDrinker) [inline concrete-core/src/backends/cuda/implementation/engines/cuda_engine/glwe_ciphertext_vector_conversion.rs:546]: no remainder management ?

File: concrete-core/src/backends/cuda/implementation/engines/cuda_engine/glwe_ciphertext_vector_discarding_conversion.rs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    /// #
+    /// # Ok(())
+    /// # }
+    /// ```
+    fn discard_convert_glwe_ciphertext_vector(
+        &mut self,
+        output: &mut GlweCiphertextVectorMutView64,
+        input: &CudaGlweCiphertextVector64,
+    ) -> Result<(), GlweCiphertextVectorDiscardingConversionError<CudaError>> {
+        unsafe { self.discard_convert_glwe_ciphertext_vector_unchecked(output, input) };
```
Comment (IceTDrinker) [inline concrete-core/src/backends/cuda/implementation/engines/cuda_engine/glwe_ciphertext_vector_discarding_conversion.rs:101]: no size checks ?

File: concrete-core/src/backends/cuda/implementation/engines/cuda_engine/glwe_ciphertext_vector_discarding_conversion.rs [hunk 5]
```diff
@@ -0,0 +1,115 @@
+use crate::backends::cuda::engines::{CudaEngine, CudaError};
+use crate::backends::cuda::implementation::entities::CudaGlweCiphertextVector64;
+use crate::prelude::GlweCiphertextVectorMutView64;
+use crate::specification::engines::{
+    GlweCiphertextVectorDiscardingConversionEngine, GlweCiphertextVectorDiscardingConversionError,
+};
+
+impl From<CudaError> for GlweCiphertextVectorDiscardingConversionError<CudaError> {
```
Comment (IceTDrinker) [inline concrete-core/src/backends/cuda/implementation/engines/cuda_engine/glwe_ciphertext_vector_discarding_conversion.rs:8]: no 32 bits ?

File: concrete-core/src/backends/default/implementation/entities/glwe_ciphertext_vector.rs [hunk 6]
```diff
@@ -64,3 +64,69 @@ pub(crate) enum GlweCiphertextVector64Version {
...
+    }
+}
+
+#[cfg(feature = "backend_default_serialization")]
+#[derive(Serialize, Deserialize)]
+pub(crate) enum GlweCiphertextVectorView64Version {
+    V0,
+    #[serde(other)]
+    Unsupported,
+}
```
Comment (IceTDrinker) [inline concrete-core/src/backends/default/implementation/entities/glwe_ciphertext_vector.rs:99]: no need for serialization version from what I can see from concrete-core main on other view types

File: concrete-core/src/backends/default/implementation/entities/glwe_ciphertext_vector.rs [hunk 7]
```diff
@@ -64,3 +64,69 @@ pub(crate) enum GlweCiphertextVector64Version {
...
+    }
+}
+
+#[cfg(feature = "backend_default_serialization")]
+#[derive(Serialize, Deserialize)]
+pub(crate) enum GlweCiphertextVectorMutView64Version {
+    V0,
+    #[serde(other)]
+    Unsupported,
+}
```
Comment (IceTDrinker) [inline concrete-core/src/backends/default/implementation/entities/glwe_ciphertext_vector.rs:132]: same

File: concrete-core/src/backends/default/implementation/entities/glwe_ciphertext_vector.rs [hunk 8]
```diff
@@ -64,3 +64,69 @@ pub(crate) enum GlweCiphertextVector64Version {
...
+#[cfg(feature = "backend_default_serialization")]
+#[derive(Serialize, Deserialize)]
+pub(crate) enum GlweCiphertextVectorView64Version {
+    V0,
+    #[serde(other)]
+    Unsupported,
+}
+
+/// A structure representing a vector of GLWE ciphertexts views with 64 bits of precision.
+/// Immutable version.
```
Comment (IceTDrinker) [inline concrete-core/src/backends/default/implementation/entities/glwe_ciphertext_vector.rs:102]: Mutable

File: concrete-core/src/backends/default/implementation/entities/glwe_ciphertext_vector.rs [hunk 9]
```diff
@@ -64,3 +64,69 @@ pub(crate) enum GlweCiphertextVector64Version {
     #[serde(other)]
     Unsupported,
 }
+
+/// A structure representing a vector of GLWE ciphertexts views with 64 bits of precision.
+/// Mutable version.
```
Comment (IceTDrinker) [inline concrete-core/src/backends/default/implementation/entities/glwe_ciphertext_vector.rs:69]: Immutable

File: concrete-core/src/backends/default/implementation/engines/default_engine/glwe_ciphertext_vector_consuming_retrieval.rs [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+use crate::backends::default::implementation::entities::{
+    GlweCiphertextVectorMutView64, GlweCiphertextVectorView64,
+};
+use crate::commons::math::tensor::IntoTensor;
+use crate::prelude::GlweCiphertextVector64;
+use crate::specification::engines::{
+    GlweCiphertextVectorConsumingRetrievalEngine, GlweCiphertextVectorConsumingRetrievalError,
+};
+
+/// # Description:
```
Comment (IceTDrinker) [inline concrete-core/src/backends/default/implementation/engines/default_engine/glwe_ciphertext_vector_consuming_retrieval.rs:11]: No 32 bits ?

File: concrete-core/src/backends/default/implementation/engines/default_engine/glwe_ciphertext_vector_creation.rs [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+    ///     polynomial_size,
+    /// )?;
+    /// #
+    /// # Ok(())
+    /// # }
+    /// ```
+    fn create_glwe_ciphertext_vector(
+        &mut self,
+        container: Vec<u64>,
+        rlwe_dimension: GlweDimension,
```
Comment (IceTDrinker) [inline concrete-core/src/backends/default/implementation/engines/default_engine/glwe_ciphertext_vector_creation.rs:49]: glwe_dimension ?

File: concrete-core/src/backends/default/implementation/engines/default_engine/glwe_ciphertext_vector_creation.rs [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+    /// ```
+    /// use concrete_commons::parameters::PolynomialSize;
+    /// use concrete_core::prelude::*;
+    /// # use std::error::Error;
+    ///
+    /// # fn main() -> Result<(), Box<dyn Error>> {
+    /// // Here we create a container outside of the engine
+    /// // Note that the size here is just for demonstration purposes and should not be chosen
+    /// // without proper security analysis for production
+    /// let glwe_size = GlweSize(600);
```
Comment (IceTDrinker) [inline concrete-core/src/backends/default/implementation/engines/default_engine/glwe_ciphertext_vector_creation.rs:26]: in practice crypto parameters rarely go above GlweDimension(2)

File: concrete-core/src/backends/default/implementation/engines/default_engine/glwe_ciphertext_vector_consuming_retrieval.rs [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+    /// // You have to make sure you size the container properly
+    /// let mut owned_container = vec![0_u64; glwe_size.0 * polynomial_size.0];
+    ///
+    /// let slice = &owned_container[..];
+    ///
+    /// // Here we just give it 0, which is totally unsafe.
+    /// const UNSAFE_SECRET: u128 = 0;
+    /// let mut engine = DefaultEngine::new(Box::new(UnixSeeder::new(UNSAFE_SECRET)))?;
+    /// let ciphertext: GlweCiphertextVector64 = engine.create_glwe_ciphertext_vector(
+    ///     slice.to_vec(),
```
Comment (IceTDrinker) [inline concrete-core/src/backends/default/implementation/engines/default_engine/glwe_ciphertext_vector_consuming_retrieval.rs:39]: I don't think this illustrates the fact we can retrieve the actual original container you can check an example from lwe_ciphertext_consuming_retrieval.rs

File: concrete-core/src/specification/engines/glwe_ciphertext_vector_consuming_retrieval.rs [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+/// A trait for engines retrieving the content of the container from a GLWE ciphertext
+/// vector consuming it in the process.
+///
+/// # Semantics
+///
+/// This [pure](super#operation-semantics) operation retrieves the content of the container from the
+/// `input` GLWE ciphertext vector consuming it in the process.
+pub trait GlweCiphertextVectorConsumingRetrievalEngine<Ciphertext, Container>:
```
Comment (IceTDrinker) [inline concrete-core/src/specification/engines/glwe_ciphertext_vector_consuming_retrieval.rs:16]: type parameters name could be improved, CiphertextVector e.g. ?

**Review by IceTDrinker — COMMENTED** [review]
Different set of comments:

Some doc string don't illustrate the retrieval like other view retrieval engines
Some crypto parameters are too big (GlweDimension generally <= 2)
Some serialization stuff as well

I would combine my reviews from the two PRs and check comments for both, as I saw different things both times

**Comment by agnesLeroy** [comment]
@IceTDrinker I need to fix the last issues here, but otherwise I tried to take all comments + the ones from the other PR into account. I may have missed some things though so a check will be welcome.

**Comment by agnesLeroy** [comment]
@slab-ci cpu_test

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

**Comment by agnesLeroy** [comment]
@slab-ci cpu_test

**Review by IceTDrinker — COMMENTED** [review]
Nearly there, if you want we can merge this so that it does not diverge, the fixes are mostly docstrings/typos/copy paste errors, so fine to fix later IMO

Plus I don't manage to read those PRs anymore, too much green text 😵

File: concrete-core/src/backends/cuda/implementation/engines/cuda_engine/glwe_ciphertext_vector_conversion.rs [hunk 15]
```diff
@@ -340,3 +335,407 @@ impl GlweCiphertextVectorConversionEngine<CudaGlweCiphertextVector64, GlweCipher
...
+    > for CudaEngine
+{
+    /// # Example
+    /// ```
+    /// use concrete_core::prelude::*;
+    /// # use std::error::Error;
+    ///
+    /// # fn main() -> Result<(), Box<dyn Error>> {
+    /// // DISCLAIMER: the parameters used here are only for test purpose, and are not secure.
+    /// use std::task::Poll;
```
[the remark this answers is not in the document]
Comment (agnesLeroy) [reply concrete-core/src/backends/cuda/implementation/engines/cuda_engine/glwe_ciphertext_vector_conversion.rs:542]: No idea, removing it

[the remark this answers is not in the document]
Comment (IceTDrinker) [reply concrete-core/src/backends/cuda/implementation/engines/cuda_engine/glwe_ciphertext_vector_conversion.rs:542]: there are a few in most of the docstrings I saw

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

**Comment by agnesLeroy** [comment]
@IceTDrinker everything should be fixed :slightly_smiling_face:

**Pull request merged**